### PR TITLE
Make daemon proxy settings configurable through daemon.json

### LINKED ISSUE #24758: Feature Request:  Enable HTTP_PROXY Configuration as part of deamon json options
Currently you can utilize GOLang's support for HTTP_PROXY/HTTPS_PROXY environment variables but they must be configured in the session/context that the daemon is launched under.

It would be valuable to be able to configure these as part of the daemon JSON such that you can deploy all of your engine's with the same configuration without needing additional configuration scripts or systemd/windows configuration.

One approach for this might be to read the values in at launch time and set the appropriate locally scoped environment variables or there may be other GoLang support for these options as this point.

### PR DESCRIPTION
**Pull request by aiordache**
fixes https://github.com/moby/moby/issues/24758

Added proxy properties to daemon configuration (config file and dockerd flags).

**- How to verify it**

```
$ cat daemon.json 
{
    "http-proxy": "http://proxytest.example.com:80",
    "https-proxy": "https://proxytest.example.com:443"
}

$ dockerd --config-file daemon.json
```

```
$ docker pull busybox
Using default tag: latest
Error response from daemon: Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp: lookup proxytest.example.com on 127.0.0.11:53: no such host


# docker build .     
Sending build context to Docker daemon  89.28MB
Step 1/3 : FROM golang:1.16-alpine AS base
Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp: lookup proxytest.example.com on 127.0.0.11:53: no such host

```

### CODE REVIEW

File: daemon/info.go [hunk 1]
```diff
@@ -64,9 +64,9 @@ func (daemon *Daemon) SystemInfo() *types.Info {
 		Labels:             daemon.configStore.Labels,
 		ExperimentalBuild:  daemon.configStore.Experimental,
 		ServerVersion:      dockerversion.Version,
-		HTTPProxy:          maskCredentials(getEnvAny("HTTP_PROXY", "http_proxy")),
-		HTTPSProxy:         maskCredentials(getEnvAny("HTTPS_PROXY", "https_proxy")),
-		NoProxy:            getEnvAny("NO_PROXY", "no_proxy"),
+		HTTPProxy:          maskCredentials(getEnvOrConfig(daemon.configStore.HttpProxy, "HTTP_PROXY", "http_proxy")),
+		HTTPSProxy:         maskCredentials(getEnvOrConfig(daemon.configStore.HttpsProxy, "HTTPS_PROXY", "https_proxy")),
+		NoProxy:            getEnvOrConfig(daemon.configStore.NoProxy, "NO_PROXY", "no_proxy"),
```
Comment (thaJeztah) [inline daemon/info.go:69]: Hm.. but this only changes what's shown in `docker info`, but not actually configures the daemon to use these proxies? 🤔 

FWIW, default behavior; https://github.com/golang/go/blob/912f0750472dd4f674b69ca1616bfaf377af1805/src/net/http/transport.go#L422-L441 / https://github.com/golang/go/blob/912f0750472dd4f674b69ca1616bfaf377af1805/src/net/http/transport.go#L43-L54

(and we are using https://github.com/docker/go-connections for some central/default configuration, which may be relevant)

File: daemon/info.go [hunk 2]
```diff
@@ -308,3 +308,18 @@ func getEnvAny(names ...string) string {
...
+	val := getEnvAny(env...)
+	if val != "" {
+		return val
+	}
+
+	if config == "" {
+		return ""
+	}
+	for _, e := range env {
+		os.Setenv(e, config)
```
Comment (thaJeztah) [inline daemon/info.go:322]: Ah, right. Guess this is the "magic" bit

**Review by ghost — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by tianon** [comment]
Honestly having a really hard time understanding the justification here -- why not use the frankly pretty well standardized environment variables instead?

**Review by fredericdalleau — APPROVED** [review]
LGTM

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Sorry for the late reply on this one. I started a review, and had a branch with changes, but didn't post it, and had to rebase my branch to still work against master. Let me know if you want to take my branch or if I should push it as a separate PR (I left some commits un-squashed to explain the steps, and should probably squash those); https://github.com/moby/moby/compare/master...thaJeztah:proxy_daemon_config_carry?expand=1

**Comment by tianon** [comment]
> Honestly having a really hard time understanding the justification here -- why not use the frankly pretty well standardized environment variables instead?

To follow up on this thought, the answer is Windows, where adding a variable to a service is anything but straightforward. 😬

**Comment by thaJeztah** [comment]
Carrying these changes with my review suggestions in https://github.com/moby/moby/pull/42835

**Pull request closed without merging**